readme.md
# handlr
[http://b23.be/handlr](http://b23.be/handlr)

## What it does

handlr is a tool for batch opening URLs.  It sits in your systray and finds any URLs in your clipboard, presents them all, and gives the ability to send some or all of them to the default browser.

Firefox and Chrome users can also copy a segment of a webpage, and handlr will find any URLs contained within the HTML.  Any other application (such as Thunderbird) which uses the same HTML clipboard format will work equally well.

handlr is written in Lua, using wxWidgets via [wxLua](http://wxlua.sourceforge.net).  A Windows build is available in the Downloads section.

## What it doesn't do (yet)

- No Linux build (coming soon)
- No website (coming soon)
- No proper build system
- No automated tests

changelog
0.5 (2012/05/03)

++ UTF-8 support fixed (with thanks to John Labenski)
++ Added non-portable mode
++ Added MinGW-specific makefile
~~ Lua errors now output to a logfile, rather than a message box

----

0.4 (2012/04/16)

++ Totally rewritten with wxLua
++ Settings dialog added
++ Configurable options added
~~ URL parsing improved
-- 32*32 icon removed (see below)
-- Dialogs no longer show a taskbar icon
-- Removed winapi and clipboard libraries

----

0.3 (2012/03/26)

++ Notification shown when the clipboard contains no URLs
++ 32*32 icon added
~~ New build procedure; executable size reduced
~~ Code cleanup
~~ URL parsing improved
-- Statically linked winapi and clipboard libs

----

0.2.1 (2012/03/16)

~~ Fixed a bug which caused handlr to not exit cleanly
~~ Minor bugfixes
-- Executable size reduced

----

0.2 (2012/03/10)

++ Firefox/Chrome's HTML clipboard format is now parsed
++ Visual Styles are now used under Windows
~~ cmd.exe window no longer appears when opening URLs

----

0.1.1 (2012/01/20)

~~ Minor bugfixes

----

0.1 (2012/01/19)

++ Initial release

frmUrls.lua
local frmUrls = {
  clip = wx.wxClipboard.Get(),
  tobj = wx.wxTextDataObject(),
  urls = {},
  search = {},
  search_html = {},
  frame = nil
}

frmUrls.hfmt = wx.wxDataFormat("HTML Format")
frmUrls.hobj = wx.wxLuaDataObjectSimple(frmUrls.hfmt)
frmUrls.hobj.data = ""
function frmUrls.hobj:GetDataHere() return true, self.data end
function frmUrls.hobj:GetDataSize() return self.data:len() end
function frmUrls.hobj:SetData(d) self.data = d return true end

function frmUrls:Show()
  if not self.frame then self:Init() end
  self.lstUrls:InsertItems(self.urls, 0)
  self.frame:Show()
  self.frame:Raise()
end

function frmUrls:Init()
  self.frame = wx.wxFrame(wx.NULL,
    wx.wxID_ANY,
    "Select links to open...",
    wx.wxDefaultPosition,
    wx.wxSize( 360,300 ),
    wx.wxCAPTION + wx.wxCLOSE_BOX + wx.wxFRAME_NO_TASKBAR + wx.wxRESIZE_BORDER + wx.wxSYSTEM_MENU + wx.wxTAB_TRAVERSAL
  )
  local lframe = self.frame
  lframe:SetSizeHints( wx.wxDefaultSize, wx.wxDefaultSize )
  lframe:SetIcon(wx.wxIcon("HANDLR_ICON", wx.wxBITMAP_TYPE_ICO_RESOURCE))
  
  self.vszUrls = wx.wxBoxSizer( wx.wxVERTICAL )
  
  self.lstUrls = wx.wxListBox( lframe, wx.wxID_ANY, wx.wxDefaultPosition, wx.wxDefaultSize, {}, wx.wxLB_MULTIPLE )
  self.vszUrls:Add( self.lstUrls, 1,  bit.bor(wx.wxALL,wx.wxEXPAND), 5 )
  
  self.hszButtons = wx.wxBoxSizer( wx.wxHORIZONTAL )
  
  self.btnCancel = wx.wxButton( lframe, wx.wxID_ANY, "Cancel", wx.wxDefaultPosition, wx.wxDefaultSize, 0 )
  self.hszButtons:Add( self.btnCancel, 0, wx.wxALL, 5 )
  
  self.hszButtons:Add( 0, 0, 1, wx.wxEXPAND, 5 )
  
  self.btnSelected = wx.wxButton( lframe, wx.wxID_ANY, "Open selected", wx.wxDefaultPosition, wx.wxDefaultSize, wx.wxBU_EXACTFIT )
  self.hszButtons:Add( self.btnSelected, 0, wx.wxALL, 5 )
  
  self.btnAll = wx.wxButton( lframe, wx.wxID_ANY, "Open all", wx.wxDefaultPosition, wx.wxDefaultSize, 0 )
  self.hszButtons:Add( self.btnAll, 0, wx.wxALL, 5 )
    
  self.vszUrls:Add( self.hszButtons, 0, wx.wxEXPAND, 5 )
    
  lframe:SetSizer( self.vszUrls )
  lframe:Layout()
  
  lframe:Centre( wx.wxBOTH )

  -- events

  lframe:Connect(wx.wxEVT_CLOSE_WINDOW, 
    function(e) self:Close() end
  )

  self.btnSelected:Connect(wx.wxEVT_COMMAND_BUTTON_CLICKED,
    function(e)
      local _, sel = self.lstUrls:GetSelections()
      local urls = {}
      for _,k in pairs(sel) do table.insert(urls, self.urls[k+1]) end
      self:OpenUrls(urls)
      self:Close()
    end
  )

  self.btnAll:Connect(wx.wxEVT_COMMAND_BUTTON_CLICKED,
    function(e)
      self:OpenUrls(self.urls)
      self:Close()
    end
  )

  self.btnCancel:Connect(wx.wxEVT_COMMAND_BUTTON_CLICKED,
    function(e)
      self:Close()
    end
  )

end

-- build list of URL patterns to search clipboard for
function frmUrls:BuildSearch()
  self.search, self.search_html = {}, {}
  for _,v in pairs(cfg.prefixes) do 
    table.insert(self.search, string.insens(v).."://[^%s/$.?#]+%.[^%s\"']+")
  end
  for _,v in pairs(cfg.subdomains) do 
    table.insert(self.search, string.insens(v).."%.[^%s\"']+") 
  end
  for _,v in pairs(cfg.attrs) do 
    for _,w in pairs(self.search) do
      table.insert(self.search_html, string.insens(v).."=[\"']?("..w..")")
    end
  end
end

-- extract URLs from contents of text/HTML on the clipboard
function frmUrls:GetUrls()
  -- parse text on clipboard
  if self.clip:GetData(self.tobj) then
    local data = self.tobj:GetText()
    for _,v in pairs(self.search) do
      for url in data:gmatch(v) do table.insert(self.urls, url) end
    end
  end
  -- parse html on clipboard
  if self.clip:GetData(self.hobj) then
    for _,v in pairs(self.search_html) do
      for url in self.hobj.data:gmatch(v) do table.insert(self.urls, url) end
    end
  end
  for k,v in pairs(self.urls) do
    -- convert URLs to lowercase
    if cfg.convert == 1 then
      if not v:match("[a-z]") then self.urls[k] = v:lower() end
    elseif cfg.convert == 2 then
      local u = string.split(v, "/")
      if #u > 2 then
        for i = 1,#u-1 do u[i] = u[i]:lower() end
        self.urls[k] = table.concat(u, "/")
      end
    elseif cfg.convert == 3 then
      self.urls[k] = v:lower()
    end
    -- prepend http:// to URLs with no prefix for better dupe removal
    if not v:match("://") then self.urls[k] = "http://"..self.urls[k] end
  end
  -- remove duplicates
  self.urls = table.uniq(self.urls)
end

-- check min/max URLs in the config against self.urls, and open them
function frmUrls:OpenUrls(urls)
  if cfg.conf_urls and #urls >= cfg.conf_urls then
    local ok = wx.wxMessageBox(
      "OK to open "..#urls.." URLs?", "Confirm", wx.wxYES_NO
    )
    if ok == wx.wxNO then return end
  end
  for _,v in pairs(urls) do wx.wxLaunchDefaultBrowser(v) end
  self.urls = {}
end

-- destroy this frame
function frmUrls:Close()
  self.urls = {}
  self.frame:Destroy()
  self.frame = nil
end

return frmUrls

frmSettings.lua
local frmSettings = {
  frame = nil
}

function frmSettings:Show()
  if not self.frame then
    self:Init()
  end
  self:Load()
  self.frame:Show()
end

function frmSettings:Init()
  self.frame = wx.wxFrame(wx.NULL,
    wx.wxID_ANY,
    "Settings",
    wx.wxDefaultPosition,
    wx.wxSize( -1,-1 ),
    wx.wxCAPTION + wx.wxCLOSE_BOX + wx.wxFRAME_NO_TASKBAR + wx.wxRESIZE_BORDER + wx.wxSYSTEM_MENU + wx.wxTAB_TRAVERSAL
  )
  local lframe = self.frame
  lframe:SetSizeHints( wx.wxDefaultSize, wx.wxDefaultSize )
  lframe:SetIcon(wx.wxIcon("HANDLR_ICON", wx.wxBITMAP_TYPE_ICO_RESOURCE))
  
  self.vszSettings = wx.wxBoxSizer( wx.wxVERTICAL )
  
  self.sbsOpen = wx.wxStaticBoxSizer( wx.wxStaticBox( lframe, wx.wxID_ANY, "Opening URLs" ), wx.wxVERTICAL )
  
  self.hszAuto = wx.wxBoxSizer( wx.wxHORIZONTAL )
  
  self.chkAuto = wx.wxCheckBox( lframe, wx.wxID_ANY, "Automatically open fewer than", wx.wxDefaultPosition, wx.wxDefaultSize, 0 )
  self.chkAuto:SetValue(true)
  self.chkAuto:SetForegroundColour( wx.wxSystemSettings.GetColour( wx.wxSYS_COLOUR_WINDOWTEXT ) )
  
  self.hszAuto:Add( self.chkAuto, 0,  bit.bor(wx.wxALIGN_CENTER_VERTICAL,wx.wxLEFT), 5 )
  
  self.txtAuto = wx.wxTextCtrl( lframe, wx.wxID_ANY, "", wx.wxDefaultPosition, wx.wxSize( 32,-1 ), wx.wxTE_RIGHT )
  self.hszAuto:Add( self.txtAuto, 0,  bit.bor(wx.wxBOTTOM,wx.wxTOP), 5 )
  
  self.lblAuto = wx.wxStaticText( lframe, wx.wxID_ANY, "URLs", wx.wxDefaultPosition, wx.wxDefaultSize, 0 )
  self.lblAuto:Wrap( -1 )
  self.hszAuto:Add( self.lblAuto, 0,  bit.bor(wx.wxALIGN_CENTER_VERTICAL,wx.wxLEFT), 3 )
  
  self.sbsOpen:Add( self.hszAuto, 0,  bit.bor(wx.wxLEFT,wx.wxRIGHT), 5 )
  
  self.hszConfirm = wx.wxBoxSizer( wx.wxHORIZONTAL )
  
  self.chkConfirm = wx.wxCheckBox( lframe, wx.wxID_ANY, "Confirm before opening more than", wx.wxDefaultPosition, wx.wxDefaultSize, 0 )
  self.chkConfirm:SetValue(true)
  self.chkConfirm:SetForegroundColour( wx.wxSystemSettings.GetColour( wx.wxSYS_COLOUR_WINDOWTEXT ) )
  
  self.hszConfirm:Add( self.chkConfirm, 0,  bit.bor(wx.wxALIGN_CENTER_VERTICAL,wx.wxLEFT), 5 )
  
  self.txtConfirm = wx.wxTextCtrl( lframe, wx.wxID_ANY, "", wx.wxDefaultPosition, wx.wxSize( 32,-1 ), wx.wxTE_RIGHT )
  self.hszConfirm:Add( self.txtConfirm, 0,  bit.bor(wx.wxBOTTOM,wx.wxTOP), 5 )
  
  self.lblConfirm = wx.wxStaticText( lframe, wx.wxID_ANY, "URLs", wx.wxDefaultPosition, wx.wxDefaultSize, 0 )
  self.lblConfirm:Wrap( -1 )
  self.hszConfirm:Add( self.lblConfirm, 0,  bit.bor(wx.wxALIGN_CENTER_VERTICAL,wx.wxLEFT), 3 )
  
  self.sbsOpen:Add( self.hszConfirm, 0,  bit.bor(wx.wxEXPAND,wx.wxLEFT,wx.wxRIGHT), 5 )
    
  self.vszSettings:Add( self.sbsOpen, 0,  bit.bor(wx.wxALL,wx.wxEXPAND), 5 )
  
  self.sbsSearch = wx.wxStaticBoxSizer( wx.wxStaticBox( lframe, wx.wxID_ANY, "Searching for URLs" ), wx.wxVERTICAL )
  
  self.fgsSearch = wx.wxFlexGridSizer( 0, 2, 0, 0 )
  self.fgsSearch:SetFlexibleDirection( wx.wxBOTH )
  self.fgsSearch:SetNonFlexibleGrowMode( wx.wxFLEX_GROWMODE_SPECIFIED )
  
  self.lblPrefixes = wx.wxStaticText( lframe, wx.wxID_ANY, "URL prefixes to search for:", wx.wxDefaultPosition, wx.wxDefaultSize, 0 )
  self.lblPrefixes:Wrap( -1 )
  self.fgsSearch:Add( self.lblPrefixes, 0,  bit.bor(wx.wxALIGN_CENTER_VERTICAL,wx.wxALL), 5 )
  
  self.txtPrefixes = wx.wxTextCtrl( lframe, wx.wxID_ANY, "", wx.wxDefaultPosition, wx.wxDefaultSize, 0 )
  self.fgsSearch:Add( self.txtPrefixes, 0,  bit.bor(wx.wxALL,wx.wxEXPAND), 5 )
  
  self.lblSubdomains = wx.wxStaticText( lframe, wx.wxID_ANY, "Subdomains to search for:", wx.wxDefaultPosition, wx.wxDefaultSize, 0 )
  self.lblSubdomains:Wrap( -1 )
  self.fgsSearch:Add( self.lblSubdomains, 0,  bit.bor(wx.wxALIGN_CENTER_VERTICAL,wx.wxALL), 5 )
  
  self.txtSubdomains = wx.wxTextCtrl( lframe, wx.wxID_ANY, "", wx.wxDefaultPosition, wx.wxDefaultSize, 0 )
  self.fgsSearch:Add( self.txtSubdomains, 0,  bit.bor(wx.wxALL,wx.wxEXPAND), 5 )
  
  self.lblAttrs = wx.wxStaticText( lframe, wx.wxID_ANY, "HTML attributes to parse:", wx.wxDefaultPosition, wx.wxDefaultSize, 0 )
  self.lblAttrs:Wrap( -1 )
  self.fgsSearch:Add( self.lblAttrs, 0,  bit.bor(wx.wxALIGN_CENTER_VERTICAL,wx.wxALL), 5 )
  
  self.txtAttrs = wx.wxTextCtrl( lframe, wx.wxID_ANY, "", wx.wxDefaultPosition, wx.wxDefaultSize, 0 )
  self.fgsSearch:Add( self.txtAttrs, 0,  bit.bor(wx.wxALL,wx.wxEXPAND), 5 )
  
  self.lblLowercase = wx.wxStaticText( lframe, wx.wxID_ANY, "Convert URLs to lowercase:", wx.wxDefaultPosition, wx.wxDefaultSize, 0 )
  self.lblLowercase:Wrap( -1 )
  self.fgsSearch:Add( self.lblLowercase, 0,  bit.bor(wx.wxALIGN_CENTER_VERTICAL,wx.wxALL), 5 )
  
  self.chcLowercaseChoices = {"Never", "If all uppercase", "Always, except filename", "Always"}
  self.chcLowercase = wx.wxChoice( lframe, wx.wxID_ANY, wx.wxDefaultPosition, wx.wxDefaultSize, self.chcLowercaseChoices, 0 )
  self.chcLowercase:SetSelection( 0 )
  self.fgsSearch:Add( self.chcLowercase, 0, wx.wxALL, 5 )
    
  self.sbsSearch:Add( self.fgsSearch, 0,  bit.bor(wx.wxEXPAND,wx.wxLEFT,wx.wxRIGHT), 5 )
    
  self.vszSettings:Add( self.sbsSearch, 0,  bit.bor(wx.wxALL,wx.wxEXPAND), 5 )
  
  self.hszButtons = wx.wxBoxSizer( wx.wxHORIZONTAL )
    
  self.hszButtons:Add( 0, 0, 1, wx.wxEXPAND, 5 )
  
  self.btnSave = wx.wxButton( lframe, wx.wxID_ANY, "Save", wx.wxDefaultPosition, wx.wxDefaultSize, 0 )
  self.hszButtons:Add( self.btnSave, 0, wx.wxALL, 5 )
  
  self.btnCancel = wx.wxButton( lframe, wx.wxID_ANY, "Cancel", wx.wxDefaultPosition, wx.wxDefaultSize, 0 )
  self.hszButtons:Add( self.btnCancel, 0, wx.wxALL, 5 )
    
  self.vszSettings:Add( self.hszButtons, 1, wx.wxEXPAND, 5 )
  
  lframe:SetSizer( self.vszSettings )
  lframe:Layout()
  self.vszSettings:Fit( lframe )
  
  lframe:Centre( wx.wxBOTH )

  -- events

  for _,v in pairs({"Auto", "Confirm"}) do
    self["chk"..v]:Connect(wx.wxEVT_COMMAND_CHECKBOX_CLICKED,
      function(e)
        if e:IsChecked() then
          self["txt"..v]:Enable(true) 
        else 
          self["txt"..v]:Disable()
        end
      end
    )

    self["txt"..v]:Connect(wx.wxEVT_COMMAND_TEXT_UPDATED,
      function(e)
        self["txt"..v]:ChangeValue(self["txt"..v]:GetValue():gsub("[^%d]", ""))
      end
    )
  end

  self.btnSave:Connect(wx.wxEVT_COMMAND_BUTTON_CLICKED, 
    function(e)
      self:Save()
      self.frame:Destroy()
      self.frame = nil
    end
  )

  self.btnCancel:Connect(wx.wxEVT_COMMAND_BUTTON_CLICKED, 
    function(e)
      self.frame:Destroy()
      self.frame = nil
    end
  )

  lframe:Connect(wx.wxEVT_CLOSE_WINDOW, 
    function(e) 
      self.frame:Destroy()
      self.frame = nil
    end
  )

end

function frmSettings:Load()
  local evt = wx.wxCommandEvent(wx.wxEVT_COMMAND_CHECKBOX_CLICKED)
  if not cfg.auto_urls then
    self.chkAuto:SetValue(false)
    wx.wxPostEvent(self.chkAuto, evt)
  else self.txtAuto:SetValue(tostring(cfg.auto_urls)) end
  if not cfg.conf_urls then
    self.chkConfirm:SetValue(false)
    wx.wxPostEvent(self.chkConfirm, evt)
  else self.txtConfirm:SetValue(tostring(cfg.conf_urls)) end
  for k,v in pairs({"Prefixes", "Subdomains", "Attrs"}) do
    self["txt"..v]:SetValue(table.concat(cfg[v:lower()], " "))
  end
  self.chcLowercase:SetSelection(cfg.convert)
end

function frmSettings:Save()
  local new = {}
  if not self.chkAuto:GetValue() or self.chkAuto:GetValue() == "" then 
    new.auto_urls = false
  else 
    new.auto_urls = tonumber(self.txtAuto:GetValue())
  end
  if not self.chkConfirm:GetValue() or self.chkConfirm:GetValue() == "" then
    new.conf_urls = false
  else 
    new.conf_urls = tonumber(self.txtConfirm:GetValue())
  end
  for k,v in pairs({"Prefixes", "Subdomains", "Attrs"}) do
    new[v:lower()] = string.split(self["txt"..v]:GetValue(), " ")
  end
  new.convert = self.chcLowercase:GetSelection()
  cfg:update(new)
  cfg:save()
end

return frmSettings

main.lua
-- handlr 0.5 (2012/05/03)
-- http://b23.be/handlr
-- by strawman at b23 dot be

-- TODO:
-- ~~ use wxConfig instead of pl.pretty?
-- ++ write tests

function table.contains(tbl, val)
  for _,v in pairs(tbl) do
    if v == val then return true end
  end
  return false
end

function table.uniq(tbl)
  local new = {}
  for _,v in pairs(tbl) do
    if not table.contains(new, v) then table.insert(new, v) end
  end
  return new
end

function string.split(str, dlm)
  new = {}
  for s in str:gmatch("[^"..dlm.."]+") do table.insert(new,s) end
  return new
end

function string.insens(str)
  return str.gsub(str, "%a", function(c)
    return(string.format("[%s%s]", c:lower(), c:upper()))
  end)
end

--

require("wx")
local frmUrls     = require("frmUrls")
local frmSettings = require("frmSettings")
local pp          = require("pl.pretty")

-- config

cfg = {
  auto_urls  = 2,
  conf_urls  = 10,
  prefixes   = {"http", "https"},
  subdomains = {"www"},
  attrs      = {"href", "src"},
  convert    = 0
}

setmetatable(cfg, {__index = {

  file = "",

  dir = function()
    return H_DIR and H_DIR or wx.wxStandardPaths.Get():GetLocalDataDir()
  end,

  update = function(self,tbl)
    for k,v in pairs(tbl) do self[k] = v end
  end,

  load = function(self, file)
    if not file then file = self.dir() .. "/handlr.cfg" end
    getmetatable(self).__index.file = file
    local cnf = io.open(file, "r")
    if not cnf then return end
    local str = cnf:read("*all")
    cnf:close()
    local tbl = pp.read(str)
    if not tbl then return end
    self:update(tbl)
  end,

  save = function(self, file)
    if not file then file = self.file end
    pp.dump(self, file)
    frmUrls:BuildSearch()
  end

}})

-- tray icon

local trayicon = wx.wxTaskBarIcon()
function trayicon:init()
  -- vv make this platform independent vv
  self:SetIcon(wx.wxIcon("HANDLR_ICON", wx.wxBITMAP_TYPE_ICO_RESOURCE), "handlr")

  self.menu = wx.wxMenu()
  self.menu_opts  = wx.wxMenuItem(self.menu, wx.wxID_ANY, "Settings")
  self.menu_about = wx.wxMenuItem(self.menu, wx.wxID_ANY, "About")
  self.menu_quit  = wx.wxMenuItem(self.menu, wx.wxID_ANY, "Exit")
  self.menu:Append(self.menu_opts)
  self.menu:AppendSeparator()
  self.menu:Append(self.menu_about)
  self.menu:Append(self.menu_quit)

  self.none = wx.wxMenu()
  self.none:Append(wx.wxMenuItem(self.none, wx.wxID_ANY, "No URLs on clipboard."))

  self:Connect(wx.wxEVT_TASKBAR_RIGHT_DOWN, 
    function(e)
      self.PopupMenu(self, self.menu)
    end
  )

  self:Connect(wx.wxEVT_TASKBAR_LEFT_DOWN,
    function(e) 
      frmUrls:GetUrls()
      if #frmUrls.urls == 0 then
        self.PopupMenu(self, self.none)
        return
      end
      if cfg.auto_urls and #frmUrls.urls <= cfg.auto_urls then
        frmUrls:OpenUrls(frmUrls.urls)
        return
      end
      frmUrls:Show()
    end
  )

  self:Connect(self.menu_opts:GetId(), wx.wxEVT_COMMAND_MENU_SELECTED,
    function(e) frmSettings:Show() end
  )

  self:Connect(self.menu_about:GetId(), wx.wxEVT_COMMAND_MENU_SELECTED,
    function(e) wx.wxLaunchDefaultBrowser("http://b23.be/handlr") end
  )

  self:Connect(self.menu_quit:GetId(), wx.wxEVT_COMMAND_MENU_SELECTED,
    function(e) print("he is") wx.wxGetApp():ExitMainLoop() end
  )

end

-- let's go

cfg:load()
frmUrls:BuildSearch()
trayicon:init()
wx.wxGetApp():MainLoop()